xrdp/xrdp_egfx.c

```diff
@@ -0,0 +1,1021 @@
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_frame_end:");
+    make_stream(s);
+    init_stream(s, 8192);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE0); /* descriptor = SINGLE */
+    /* RDP8_BULK_ENCODED_DATA */
+    out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
+    /* RDPGFX_HEADER */
+    out_uint16_le(s, XR_RDPGFX_CMDID_ENDFRAME); /* cmdId */
+    out_uint16_le(s, 0); /* flags = 0 */
+    holdp = s->p;
+    out_uint8s(s, 4); /* pduLength, set later */
+    out_uint32_le(s, frame_id);
+    s_mark_end(s);
+    bytes = (int) ((s->end - holdp) + 4);
+    s->p = holdp;
+    out_uint32_le(s, bytes);
+    bytes = (int) (s->end - s->data);
+    error = xrdp_egfx_send_data(egfx, s->data, bytes);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_send_frame_end: "
+        " xrdp_egfx_send_data error %d", error);
+    free_stream(s);
+    return error;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_capsconfirm(struct xrdp_egfx *egfx, int version, int flags)
+{
+    int error;
+    int bytes;
+    struct stream *s;
+    char *holdp;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_capsconfirm:");
+    make_stream(s);
+    init_stream(s, 8192);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE0); /* descriptor = SINGLE */
+    /* RDP8_BULK_ENCODED_DATA */
+    out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
+    /* RDPGFX_HEADER */
+    out_uint16_le(s, XR_RDPGFX_CMDID_CAPSCONFIRM); /* cmdId */
+    out_uint16_le(s, 0); /* flags = 0 */
+    holdp = s->p;
+    out_uint8s(s, 4); /* pduLength, set later */
+    out_uint32_le(s, version); /* version */
+    out_uint32_le(s, 4); /* capsDataLength */
+    out_uint32_le(s, flags);
+    s_mark_end(s);
+    bytes = (int) ((s->end - holdp) + 4);
+    s->p = holdp;
+    out_uint32_le(s, bytes);
+    bytes = (int) (s->end - s->data);
+    error = xrdp_egfx_send_data(egfx, s->data, bytes);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_send_capsconfirm: "
+        " xrdp_egfx_send_data error %d", error);
+    free_stream(s);
+    return error;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_wire_to_surface1(struct xrdp_egfx *egfx, int surface_id,
+                                int codec_id, int pixel_format,
+                                struct xrdp_egfx_rect *dest_rect,
+                                void *bitmap_data, int bitmap_data_length)
+{
+    int error;
+    int bytes;
+    int index;
+    int segment_size;
+    int segment_count;
+    struct stream *s;
+    char *hold_segment_count;
+    char *bitmap_data8;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_wire_to_surface1:");
+    make_stream(s);
+    bytes = bitmap_data_length + 8192;
+    bytes += 5 * (bitmap_data_length / 0xFFFF);
+    init_stream(s, bytes);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE1); /* descriptor = MULTIPART */
+    hold_segment_count = s->p;
+    out_uint8s(s, 2); /* segmentCount set later */
```

Comment:
You will run into problems here:

First of all, `MULTIPART` is not always necessary, especially with AVC, but sometimes it is. So, you waste bandwidth here.

But this is not the problem. The problem is that you assume the `segmentCount` to be 2. If you don't need more, this is fine, even for tiny frame updates.
But if you have an update `> 2 * 65535` Bytes, you have a problem, because you will need more segments.
People with large display setup will run into this, or when AVC is not available and you fall back to Calista Progressive.

I suggest you to handle the segmentation outside this file and apply it in a general way.

Similar situation for `WireToSurface_2` and the `Fill` PDU above (and actually also for the blit PDUs (CacheToSurface, SurfaceToSurface)).

Reply:
Ah, nvm. Just see that you actually skip these Bytes here and set the `segmentCount` afterwards.

Reply:
Resolving.